cranelift/codegen/src/ir/pcc.rs

```diff
@@ -0,0 +1,347 @@
+            }
+            _ => false,
+        }
+    }
+
+    /// Computes whatever fact can be known about the sum of two other
+    /// values with attached facts. The add is performed to the given
+    /// bit-width, and the target machine has the given pointer-width.
+    pub fn add(&self, other: &Self, add_width: u16, pointer_width: u16) -> Option<Self> {
+        match (self, other) {
+            (
+                Fact::ValueMax {
+                    bit_width: bw_lhs,
+                    max: lhs,
+                },
+                Fact::ValueMax {
+                    bit_width: bw_rhs,
+                    max: rhs,
+                },
+            ) if bw_lhs == bw_rhs && add_width >= *bw_lhs => {
+                let computed_max = lhs.checked_add(*rhs)?;
+                Some(Fact::ValueMax {
+                    bit_width: *bw_lhs,
+                    max: computed_max,
+                })
+            }
+
+            (
+                Fact::ValueMax {
+                    bit_width: bw_max,
+                    max,
+                },
+                Fact::PointsTo { region },
+            )
+            | (
+                Fact::PointsTo { region },
+                Fact::ValueMax {
+                    bit_width: bw_max,
+                    max,
+                },
+            ) if *bw_max >= pointer_width && add_width >= *bw_max => {
+                let computed_region = MemoryRegion {
+                    max: region.max.checked_sub(*max)?,
+                };
+                Some(Fact::PointsTo {
+                    region: computed_region,
+                })
+            }
+
+            _ => None,
+        }
+    }
+
+    /// Computes the `uextend` of a value with the given facts.
+    pub fn uextend(&self, from_width: u16, to_width: u16) -> Option<Self> {
+        match self {
+            // If we have a defined value in bits 0..bit_width, and we
+            // are filling zeroes into from_bits..to_bits, and
+            // bit_width and from_bits are exactly contiguous, then we
+            // have defined values in 0..to_bits (and because this is
+            // a zero-extend, the max value is the same).
+            Fact::ValueMax { bit_width, max } if *bit_width == from_width => Some(Fact::ValueMax {
+                bit_width: to_width,
+                max: *max,
+            }),
+            _ => None,
+        }
+    }
+
+    /// Computes the `sextend` of a value with the given facts.
+    pub fn sextend(&self, from_width: u16, to_width: u16) -> Option<Self> {
+        match self {
+            // If we have a defined value in bits 0..bit_width, and
+            // the MSB w.r.t. `from_width` is *not* set, then we can
+            // do the same as `uextend`.
+            Fact::ValueMax { bit_width, max }
+                if *bit_width == from_width && (*max & (1 << (*bit_width - 1)) == 0) =>
+            {
+                self.uextend(from_width, to_width)
+            }
```

Comment:
And otherwise we could set it to `bit_width::MAX` -- any reason we shouldn't do that and returning `None` instead is preferred?

Reply:
Yep, good point, added.